src/app/components/import/ImportModal.js

```diff
@@ -0,0 +1,207 @@
+import PropTypes from 'prop-types';
+import { c } from 'ttag';
+import {
+    useNotifications,
+    useUser,
+    useUserKeys,
+    useModals,
+    FormModal,
+    ConfirmModal,
+    Alert,
+    ResetButton,
+    PrimaryButton
+} from 'react-components';
+
+import AttachingModalContent from './AttachingModalContent';
+import ImportCsvModalContent from './ImportCsvModalContent';
+import ImportingModalContent from './ImportingModalContent';
+import ImportGroupsModalContent from './ImportGroupsModalContent';
+
+import { hasCategories } from '../../helpers/import';
+import { readCsv } from '../../helpers/csv';
+import { readVcf } from '../../helpers/vcard';
+import { BASE_SIZE } from 'proton-shared/lib/constants';
+
+const [ATTACHING, ATTACHED, CHECKING_CSV, IMPORTING, IMPORT_GROUPS] = [1, 2, 3, 4, 5];
+const MAX_SIZE = 10 * BASE_SIZE ** 2; // 10 MB
+
+const getI18nTitle = () => ({
+    [ATTACHING]: c('Title').t`Import contacts`,
+    [ATTACHED]: c('Title').t`Import contacts`,
+    [CHECKING_CSV]: c('Title').t`Import CSV file`,
+    [IMPORTING]: c('Title').t`Importing contacts`,
+    [IMPORT_GROUPS]: c(`Title`).t`Import groups`
+});
+
+const ImportModal = ({ onClose, ...rest }) => {
+    const title = getI18nTitle();
+
+    const { createModal } = useModals();
+    const { createNotification } = useNotifications();
+    const [user] = useUser();
+    const [userKeysList, loadingUserKeys] = useUserKeys(user);
+
+    const [step, setStep] = useState(ATTACHING);
+    const [file, setFile] = useState({});
+    const [vcardContacts, setVcardContacts] = useState([]);
+    const [encryptingDone, setEncryptingDone] = useState(false);
+
+    const handleClear = () => {
+        setFile({});
+        setStep(ATTACHING);
+    };
+
+    const handleAttach = ({ target }) => {
+        const attachedFile = [...target.files].filter(({ type }) => ['text/vcard', 'text/csv'].includes(type))[0];
+
+        if (target.files.length && !attachedFile) {
+            return createNotification({
+                type: 'error',
+                text: c('Error notification').t`No .csv or .vcard file selected`
+            });
+        }
+        if (attachedFile.size > MAX_SIZE) {
+            return createModal(
+                <ConfirmModal
+                    onConfirm={handleClear}
+                    onClose={handleClear}
+                    confirm={c('Action').t`Go back`}
+                    close={null}
+                    title={c('Title').t`File is too big!`}
+                >
+                    <Alert type="error">{c('Error info')
+                        .t`We only support importing files smaller than 10 MB. Please split your contacts into several smaller files.`}</Alert>
+                </ConfirmModal>
+            );
+        }
+
+        setStep(ATTACHED);
+        setFile({ ...file, attached: attachedFile });
+    };
+
+    const handleEncryptingDone = () => setEncryptingDone(true);
+
+    const { content, ...modalProps } = (() => {
```

Comment:
This functions behaves like a switch and seems to be to complicated. From where I see it, it should be 4 child components, this component would then just select the right  component from the render function. I think it sould be more readable.

Reply:
I agree it's not so easy to read, but we need a modal that changes its content four times (there are four steps the user has to go through in the import process). We thought this was the best solution despite its ugliness.